File: src/main/java/com/github/ldaniels528/interview/javaprep/sorting/BinarySearch.java
package com.github.ldaniels528.interview.javaprep.sorting;

import java.util.Arrays;
import java.util.Random;

public class BinarySearch {

    public static void main(String[] args) {
        final Random random = new Random();
        final int[] numbers = new int[100];
        for (int n = 0; n < numbers.length; n++) numbers[n] = random.nextInt(1000);
        Arrays.sort(numbers);

        final int number = numbers[random.nextInt(numbers.length)];

        System.out.printf("Looking for value %d...\n", number);
        final int index = binarySearch(numbers, number);
        if (index == -1) {
            System.out.printf("Value %d was not found\n", number);
            System.out.println();
            Arrays.stream(numbers).forEach(n -> System.out.printf("number: %d\n", n));

        }
        else System.out.printf("Value %d was found at %d\n", number, index);
    }

    private static int binarySearch(final int[] array, final int number) {
        int p0 = 0;
        int p1 = array.length - 1;

        while (p0 < p1 && array[p0] != number && array[p1] != number) {
            final int mp = (p0 + p1) / 2;
            if (number >= array[mp]) p0 = mp;
            if (number <= array[mp]) p1 = mp;
        }

        if (array[p0] == number) return p0;
        else if (array[p1] == number) return p1;
        else return -1;
    }

}

File: src/main/scala/com/github/ldaniels528/interview/scalaprep/sorting/BinarySearch.scala
package com.github.ldaniels528.interview.scalaprep.sorting

import com.github.ldaniels528.interview.scalaprep.util.ComparableHelper._

import scala.util.Random

/**
  * Binary Search
  */
object BinarySearch extends App {

  val numbers = (0 to 100000 by 10).map(v => v: Integer)
  val number = numbers(Random.nextInt(numbers.length))

  numbers.searchFor(number) match {
    case Some(index) =>
      println(s"$number is at index $index")
    case None =>
      println(s"$number was not found")
  }

  final implicit class BinarySearchExtensions[T <: Comparable[T]](val values: Seq[T]) extends AnyVal {

    /**
      * Performs a binary search
      * Lookup complexity: O(log n)
      * @param value the value to search for
      * @return
      */
    @inline
    def searchFor(value: T): Option[Int] = {
      var p0 = 0
      var p1 = values.length - 1
      var changed = true

      while (p0 != p1 && values(p0) < value && values(p1) > value && changed) {
        val mp = (p0 + p1) / 2
        println(s"$value: mp = $mp (${values(mp)}), p0 = $p0 (${values(p0)}), p1 = $p1 (${values(p1)})")

        val z0 = p0
        val z1 = p1
        if (value >= values(mp)) p0 = mp else p1 = mp
        changed = z0 != p0 || z1 != p1
      }

      if (values(p0) == value) Some(p0)
      else if (values(p1) == value) Some(p1)
      else None
    }
  }

}
